Add toDescriptor to ResolvedMethodDeclaration

0verhead opened the pull request:
Fixes #3808.

Adds `toDescriptor()` to ResolvedMethodDeclarations with all the needed implementations. It was also necessary to add `toDescriptor()` for ResolvedType.

For convenience, a TypeUtils class was added that contains utility methods for getting the descriptors for `java.lang.reflect.Method` and primitive classes (compare to [ASM](https://gitlab.ow2.org/asm/asm/-/blob/master/asm/src/main/java/org/objectweb/asm/Type.java#L580)). 

### Found issue:
Inner classes(/enums) do not have the correct fully qualified name. You get Class.InnerClass instead of Class$InnerClass. This also effects my test cases for the `valueOf` and `values` method of enums.
For now, it should be fine, but I think this needs to be addressed in a future issue and PR.

0verhead committed:
Add toDescriptor to ResolvedMethodDeclaration
- add necessary implementations
- add toDescriptor to ResolvedType with necessary implementations
- extend DescriptorTest with new cases

jlerbsc commented on javaparser-core/src/main/java/com/github/javaparser/utils/TypeUtils.java:
```diff
@@ -0,0 +1,86 @@
+        String descriptor;
+        if (clazz == Void.TYPE) {
+            descriptor = new VoidType().toDescriptor();
+        } else if (clazz == Integer.TYPE) {
```
Maybe you can simplify this structure by using ResolvedPrimitiveType.byBoxTypeQName(..)

0verhead replied:
I think such a method only exists for ResolvedPrimitiveType and not PrimitiveType. Unfortunately, `ResolvedPrimitiveType.byBoxTypeQName` won't be useful here as it only returns Resolved(Primitive)Types.

But I can implement a similar method for PrimitiveTypes. Maybe like this:
```java
public static Optional<Primitive> getPrimitiveByBoxedTypeName(String name) {
    for (Primitive prim : Primitive.values()) {
        if (prim.nameOfBoxedType.equals(name)) {
            return Optional.of(prim);
        }
    }
    return Optional.empty();
}
```

And then in `TypeUtils.getPrimitiveTypeDescriptor` I can do:
```java
// void case:
...
// primitve cases:
Optional<Primitive> optPrim = PrimitiveType.getPrimitiveByBoxedTypeName(clazz.getCanonicalName());
if (optPrim.isPresent()) {
    return new PrimitiveType(optPrim.get()).toDescriptor();
}
```

jlerbsc replied:
> I think such a method only exists for ResolvedPrimitiveType and not PrimitiveType. Unfortunately, `ResolvedPrimitiveType.byBoxTypeQName` won't be useful here as it only returns Resolved(Primitive)Types.

You are right I meant ResolvedPrimitiveType.
Why do you say it's not useful? You can do something like this (or something better with a little more work).

    @Test
	void test() {
		Optional<ResolvedType> rt = ResolvedPrimitiveType.byBoxTypeQName(Integer.class.getCanonicalName());
		String type = toDescritor(rt.get().asPrimitive().name());
		assertEquals("I", type);
	}
	
	private String toDescritor(String primitiveTypeName) {
		return getPrimitiveType(primitiveTypeName)
				.map(p -> new PrimitiveType(p).toDescriptor())
				.orElseThrow(() -> new IllegalArgumentException("Unknown primitive"));
	}
	
	private Optional<Primitive> getPrimitiveType(String name) {
		return Optional.of(Primitive.valueOf(name));
	}

0verhead replied:
Ah, that's clever.
What are your thoughts on adding a (getter) method `toDescriptor` to `Primitive` and then
```java
public String getPrimitiveTypeDescriptor(final Class<?> clazz) {
    String className = clazz.getCanonicalName();
    Optional<Primitive> optPrim = Optional.of(Primitive.valueOf(className));
    return optPrim.map(p -> p.toDescriptor())
            .orElseThrow(() -> new IllegalArgumentException("Unknown primitive"));
}
```

Wouldn't that be even cleaner?

jlerbsc replied:
Why not but in this case it is also necessary to adapt the toDescriptor method in the PrimitiveType class.

0verhead replied:
The new commit should contain the mentioned improvements. Additionally I realized, that in my original code only for the 'real' primitive classes (int.class, byte.class etc.) the correct descriptor was returned. Passing a wrapper class (like Integer.class) into the method resulted in an error.

I fixed that. Now you should get the correct descriptor for all primitive classes as well as their wrapper classes.
The solution looks like this:
```java
public static String getPrimitiveTypeDescriptor(final Class<?> clazz) {
    if (clazz == Void.TYPE || clazz == Void.class) {
        return new VoidType().toDescriptor();
    }
    String className = clazz.getCanonicalName();
    Optional<Primitive> prim = getPrimitive(className);
    if (prim.isPresent()) {
        return prim.get().toDescriptor();
    }
    Optional<ResolvedType> rt = ResolvedPrimitiveType.byBoxTypeQName(className);
    if (rt.isPresent()) {
        return getResolvedPrimitiveTypeDescriptor(rt);
    }
    throw new IllegalArgumentException("Unknown primitive: " + className);
}

private static String getResolvedPrimitiveTypeDescriptor(Optional<ResolvedType> rt) throws AssertionError {
    String typeName = rt.get().asPrimitive().describe();
    Optional<Primitive> prim = getPrimitive(typeName);
    return prim.map(p -> p.toDescriptor())
            .orElseThrow(() -> new AssertionError(
                    String.format(
                            "ResolvedPrimitiveType name \"%s\" does not match any Primitive enum constant identifier.",
                            typeName.toUpperCase())));
}

private static Optional<Primitive> getPrimitive(String name) {
    try {
        return Optional.of(Primitive.valueOf(name.toUpperCase()));
    } catch (IllegalArgumentException e) {
        return Optional.empty();
    }
}
```

But now I am honestly wondering, if it would'nt  be better to just do:
```java
if (clazz == Void.TYPE || clazz == Void.class) {
    descriptor = new VoidType().toDescriptor();
} else if (clazz == Integer.TYPE || clazz == Integer.class) {
    descriptor = Primitive.INT.toDescriptor();
} else if (clazz == Boolean.TYPE || clazz == Boolean.class) {
    ...
```

jlerbsc commented on javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/DescriptorTest.java:
```diff
@@ -17,18 +22,35 @@
+                "  }\n" +
                 "}";
+
+        Path javassistJar = adaptPath("src/test/resources/issue3808/javassist-3.29.2-GA.jar");
```
Do not include the javassists library in the project resources

0verhead replied:
I have removed the jar and used the javassistmethoddecl.jar from the resources.

0verhead committed:
Change TypeUtils.getPrimitiveTypeDescriptor structure
- remove else if cases
- get correct descriptor for primitive classes (e.g. int.class)
- and also their wrapper classes (Integer.class)
- add toDescriptor to Primitive

jlerbsc commented on javaparser-core/src/main/java/com/github/javaparser/utils/TypeUtils.java:
```diff
@@ -59,28 +62,36 @@ private static void appendDescriptor(final Class<?> clazz, final StringBuilder s
+        if (clazz == Void.TYPE || clazz == Void.class) {
+            return new VoidType().toDescriptor();
+        }
+        String className = clazz.getCanonicalName();
```
It still seems a bit too complex to me.

If you add 2 methods in Primitive

    public static Optional<Primitive> byBoxedType(String qName) {
            for (Primitive primitive : values()) {
                if (primitive.nameOfBoxedType.equals(qName)) {
                    return Optional.of(primitive);
                }
            }
            return Optional.empty();
        }
        
        public String toDecriptor() {
        	return descriptor;
        }

then you can do it just like that.

    @Test
	void test0() {
		Optional<Primitive> primitive = Primitive.byBoxedType(Integer.class.getSimpleName());
		assertEquals("I", primitive.map(type -> type.toDecriptor()).orElseThrow(() -> new IllegalArgumentException("Unknown primitive type")));
	}

Also don't throw an AssertionException and an optional parameter is bad practice.

0verhead replied:
The newest commit https://github.com/javaparser/javaparser/pull/3819/commits/eaf1f0d06727b9262f3189c01e1f2951e4fc5bc4 basically did what you suggest using the unboxMap and also adding `Primitive.byTypeName` for all the inputs like int.class, boolean.class etc.

0verhead committed:
Simplify TypeUtils.getPrimitiveTypeDescriptor

0verhead committed:
Restructure DescriptorTest and remove dependency

0verhead committed:
Add TypeDescriptorTest, rename DescriptorTest
- TypeDescriptorTest testing descriptors for all primitive types
- rename DescriptorTest to MethodDescriptorTest

0verhead committed:
Fix import order according to checkstyle

0verhead committed:
Merge branch 'issue3808' of github.com:vanHekthor/javaparser into issue3808

jlerbsc committed:
Merge branch 'master' into issue3808
